lib/realtime/vehicle.ex

```diff
@@ -64,20 +76,122 @@ defmodule Realtime.Vehicle do
   """
   @spec decode(term()) :: t()
   def decode(%{} = json) do
+    stop_times_on_trip_fn =
+      Application.get_env(:realtime, :stop_times_on_trip_fn, &Gtfs.stop_times_on_trip/1)
+
+    id = json["id"]
+    label = json["vehicle"]["vehicle"]["label"]
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_stop_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stop_times_on_trip =
+      try do
+        stop_times_on_trip_fn.(trip_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    {percent_of_the_way_to_next_timepoint, next_timepoint_stop_time} =
+      percent_of_the_way_to_next_timepoint(stop_times_on_trip, stop_id)
+
     %__MODULE__{
-      id: json["id"],
-      label: json["vehicle"]["vehicle"]["label"],
-      timestamp: json["vehicle"]["timestamp"],
-      direction_id: json["vehicle"]["trip"]["direction_id"],
-      route_id: json["vehicle"]["trip"]["route_id"],
-      trip_id: json["vehicle"]["trip"]["trip_id"],
-      current_status: decode_current_status(json["vehicle"]["current_status"]),
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      current_stop_status: current_stop_status,
+      stop_id: stop_id,
+      current_timepoint_status:
+        timepoint_status(current_stop_status, next_timepoint_stop_time, stop_id),
+      timepoint_id: next_timepoint_id(next_timepoint_stop_time),
```

Comment:
could be `next_timepoint_stop_time && next_timepoint_stop_time.timepoint_id` instead of needing its own function.

Reply:
Dialyzer was complaining about that, that's what I had initially.